### PR TITLE
BCDA-2751: Optimize Beneficiary De-Dup Query to Use a Subquery

### PR DESCRIPTION
**Pull request by msnook**
### Fixes [BCDA-2751](https://jira.cms.gov/browse/BCDA-2751)

This is a follow-up to the first iteration of de-duplicating beneficiaries that might show up in an ACO's attribution list.  The first iteration can be found here: https://github.com/CMSgov/bcda-app/pull/506

In this follow-up iteration, the de-dup query has been optimized to use a sub-query.  This ensures that a large array is not used in the SQL computation, which could result in the following Postgres error:
```
got xxxxxx parameters but PostgreSQL only supports 65535 parameters
```

### Proposed Changes

Details on proposed guidance acquired from [here](https://stackoverflow.com/questions/52712022/what-is-the-most-performant-way-to-rewrite-a-large-in-clause).

### Change Details

Updated de-dup logic to use `id IN (subquery)` instead of `id IN (value [, ...])`

### Security Implications

The updated/refactored query handles PHI/PII data.  It has been tested thoroughly through automated regression tests and manual scenarios.

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [ ] no PHI/PII is affected by this change

### Acceptance Validation

- All existing and un-modified unit/integration tests pass, which means this query refactoring has not regressed.
- Locally, I created a scenario where an ACO had >65535 beneficiaries, and was able to use the code in its original state (prior to implementing this change) to result in the above mentioned Postgres error when a bulk request was made.  Then, I applied the changes from this PR and confirmed that a bulk request can be made successfully.

### Feedback Requested

Please review.

### CODE REVIEW

File: bcda/models/models.go [hunk 1]
```diff
@@ -322,14 +322,13 @@ func (aco *ACO) GetNewAndExistingBeneficiaries(includeSuppressed bool, since str
...
 
 		// this is used to get unique ids for de-duplicating MBIs that are listed multiple times in the CCLF8 file
-		var uniqueIds []int64
-		db.Raw("SELECT id FROM ( SELECT max(id) as id, mbi FROM cclf_beneficiaries WHERE file_id = ? GROUP BY mbi ) as id", cclfFileNew.ID).Pluck("id", &uniqueIds)
+		uniqueIdQuery := fmt.Sprintf("SELECT id FROM ( SELECT max(id) as id, mbi FROM cclf_beneficiaries where file_id = %d GROUP BY mbi ) as id", cclfFileNew.ID)
 
 		// Populate new beneficiaries collection
 		if suppressedMBIs != nil {
-			err = db.Not("mbi", suppressedMBIs).Not("mbi", cclfBeneficiariesOld).Where("id IN (?)", uniqueIds).Find(&newBeneficiaries, "file_id = ?", cclfFileNew.ID).Error
+			err = db.Not("mbi", suppressedMBIs).Not("mbi", cclfBeneficiariesOld).Where("id IN (?)", db.Raw(uniqueIdQuery).SubQuery()).Find(&newBeneficiaries, "file_id = ?", cclfFileNew.ID).Error
```
Comment (mtrang1263) [inline bcda/models/models.go:329]: Should we also consider adding this protection in the following spots?

1. Not("mbi", suppressedMBIs)
2. Not("mbi", cclfBeneficiariesOld)
3. Where("mbi IN (?)", cclfBeneficiariesOld)

Comment (msnook) [reply bcda/models/models.go:329]: Absolutely necessary.   Great catch @mtrang1263 !

Comment (msnook) [reply bcda/models/models.go:329]: The `suppressedMBIs` arrays are bound here, so I don't think there is a pressing need to refactor those.  If you think strongly, we can add a separate ticket to address.  
I definitely agree that the `cclfBeneficiariesOld` arrays should be refactored to use a SubQuery.  Great catch @mtrang1263 !  I have updated in latest commit.

File: bcda/models/models.go [hunk 2]
```diff
@@ -310,7 +309,8 @@ func (aco *ACO) GetNewAndExistingBeneficiaries(includeSuppressed bool, since str
 	// Get all the beneficiaries from the older CCLF for this ACO, to use as a basis for comparison
 	// If the older CCLF was not found, then all beneficiaries will be considered new
 	if foundCclfFileOld {
-		err = db.Table("cclf_beneficiaries").Where("file_id = ?", cclfFileOld.ID).Pluck("mbi", &cclfBeneficiariesOld).Error
+		cclfBeneficiariesOldQuery := fmt.Sprintf("SELECT mbi FROM cclf_beneficiaries WHERE file_id = %d", cclfFileOld.ID)
+		err = db.Raw(cclfBeneficiariesOldQuery).Error
```
Comment (mtrang1263) [inline bcda/models/models.go:313]: We don't need this query, right? We're just dropping the results and requerying at lines 329 and 331

Comment (msnook) [reply bcda/models/models.go:313]: I was thinking we could "error fast", like we did previously, so as not to affect existing flow.

Comment (mtrang1263) [reply bcda/models/models.go:313]: Gotcha, I can see the benefit of the early exit.

**Pull request merged**